File: src/main/java/com/skocur/imagecipher/encrypters/MultiColorEncryption.java
package com.skocur.imagecipher.encrypters;

import com.imagecipher.icsdk.annotations.IcAlgorithmSpecification;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.io.IOException;

@IcAlgorithmSpecification(algorithmName = "Multi Color Encryption")
public class MultiColorEncryption extends Encrypter {

  public MultiColorEncryption(@NotNull String fileName) throws IOException {
    super(fileName);
  }

  @Override
  public void encrypt(@NotNull String text) {
    int index = 0;
    boolean hasDone = false;

    for (int i = 0; i < image.getHeight() && !hasDone; ++i) {
      for (int j = 0; j < image.getWidth() && !hasDone; ++j) {
        if (index < text.length()) {
          Color color = encryptBlue(text.charAt(index++), j, i);

          image.setRGB(j, i, color.getRGB());
        } else {
          hasDone = true;
        }
      }
    }
  }

  @NotNull
  private Color encryptBlue(char character, int posX, int posY) {
    int argb = image.getRGB(posX, posY);

    int r = (argb >> 16) & 0b11111111;
    int g = (argb >> 8) & 0b11111111;

    int newB = 0;
    if (character < 255) {
      newB = character & 0b11111111;
    }

    return new Color(r, g, newB);
  }
}

File: app/src/main/kotlin/com/imagecipher/app/encrypters/MultiColorEncryption.kt
package com.imagecipher.app.encrypters

import com.imagecipher.app.plugin.lib.annotations.IcAlgorithmSpecification
import org.apache.logging.log4j.LogManager
import java.awt.Color

@IcAlgorithmSpecification(algorithmName = "Multi Color Encryption")
class MultiColorEncryption(fileName: String) : Encrypter(fileName) {

    companion object {
        private val logger = LogManager.getLogger()
    }

    override fun encrypt(text: String) {
        logger.debug("Encrypting: $text")
        var index = 0
        var hasDone = false

        for (i in 0 until image.height) {
            if (hasDone) break
            for (j in 0 until image.width) {
                if (index < text.length) {
                    val color = encryptBlue(text[index++], j, i)
                    image.setRGB(j, i, color.rgb)
                } else {
                    hasDone = true
                    break
                }
            }
        }
    }

    private fun encryptBlue(character: Char, posX: Int, posY: Int): Color {
        val argb = image.getRGB(posX, posY)

        val r = (argb shr 16) and 0b11111111
        val g = (argb shr 8) and 0b11111111

        val newB = if (character.code < 255) character.code and 0b11111111 else 0

        return Color(r, g, newB)
    }
}
